File: solutions/java/src/parkinglot/ParkingSpot.java
package parkinglot;

import parkinglot.VehicleType.Vehicle;
import parkinglot.VehicleType.VehicleType;

public class ParkingSpot {

    private final int spotNumber;
    private VehicleType vehicleType;
    private Vehicle parkedVehicle;

    public ParkingSpot(int spotNumber) {
        this.spotNumber = spotNumber;
    }

    public synchronized boolean isAvailable(){
        return parkedVehicle == null;
    }

    public synchronized void parkVehicle(Vehicle vehicle) {
        if(isAvailable() ){
            parkedVehicle = vehicle;
            vehicleType = vehicle.getType();
        }else{
            throw new IllegalArgumentException("Already occupied the parking spot");
        }
    }

    public synchronized void unparkVehicle() {
        parkedVehicle = null;
        vehicleType = null;
    }

    public VehicleType getVehicleType() {
        return vehicleType;
    }

    public Vehicle getParkedVehicle() {
        return parkedVehicle;
    }

    public int getSpotNumber() {
        return spotNumber;
    }

}

File: solutions/python/Parking Lot System/python/parking_spot.py
from models.vehicle import Vehicle
from models.vehicle_type import VehicleType


class ParkingSpot:
    def __init__(self, spot_number: int):
        self.spot_number = spot_number
        self.parked_vehicle = None

    def is_available(self) -> bool:
        return self.parked_vehicle is None

    def park_vehicle(self, vehicle: Vehicle):
        if self.is_available():
            self.parked_vehicle = vehicle
        else:
            raise ValueError("Spot not available")

    def unpark_vehicle(self):
        self.parked_vehicle = None

    def get_vehicle_type(self) -> VehicleType:
        return self.parked_vehicle.vehicle_type

    def get_parked_vehicle(self) -> Vehicle:
        return self.parked_vehicle

    def get_spot_number(self) -> int:
        return self.spot_number
